enclave/src/polkadex_gateway.rs

```diff
@@ -0,0 +1,137 @@
+use polkadex_sgx_primitives::types::{Order, OrderSide, OrderType, OrderUUID, TradeEvent};
+use sgx_types::{sgx_status_t, SgxResult};
+
+use crate::polkadex;
+use crate::polkadex_balance_storage;
+use crate::polkadex_orderbook_storage;
+
+/// Place order function does the following
+/// 1. authenticate
+/// 2. mutate balances (reserve amount offered in order)
+/// 3. store_order (async)
+/// 4. send order to OpenFinex API
+/// 5. report OpenFinex API result to sender
+pub fn place_order(main_account: AccountId, proxy_acc: Option<AccountId>, order: Order) -> SgxResult<()> {
+    // Authentication
+    authenticate_user(main_account.clone(), proxy_acc)?;
+    // Mutate Balances
+    match order.order_type {
+        OrderType::LIMIT => {
+            if let Some(price) = order.price {
+                match order.side {
+                    OrderSide::BID => {
+                        // TODO: Is it safe to use saturating_mul here?
+                        let amount = price.saturating_mul(order.quantity);
+                        polkadex_balance_storage::reserve_balance(&main_account, order.market_id.quote, amount)?;
+                    }
+                    OrderSide::ASK => {
+                        // TODO: Is it safe to use saturating_mul here?
+                        polkadex_balance_storage::reserve_balance(&main_account, order.market_id.base, order.quantity)?;
+                    }
+                }
+            } else {
+                error!("Price not given for a limit order");
+                return Err(sgx_status_t::SGX_ERROR_UNEXPECTED); // TODO: It would be better if we can return a error response for RPC to handle
+            }
+        }
+        OrderType::MARKET => {
+            match order.side {
+                OrderSide::BID => {
+                    // TODO: Is it safe to use saturating_mul here?
+                    // TODO: How do we reserve trade amount for Market Buy since it is not possible to define price before
+                    // TODO: order has been matched in the orderbook.
+                    // let amount = price.saturating_mul(order.quantity);
+                    // polkadex_balance_storage::reserve_balance(&main_account, order.market_id.quote, amount)?;
+                    error!("Market Buy is not implemented");
+                    return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+                }
+                OrderSide::ASK => {
+                    // TODO: Is it safe to use saturating_mul here?
+                    polkadex_balance_storage::reserve_balance(&main_account, order.market_id.base, order.quantity)?;
+                }
+            }
+        }
+        OrderType::FillOrKill | OrderType::PostOnly => {
+            error!("OrderType is not implemented");
+            return Err(sgx_status_t::SGX_ERROR_UNEXPECTED);
+        }
+    }
+    // Store the order
+    // TODO: we need OrderUUID for storing the order, it is given by Openfinex but we need OrderUUID to
```

Comment:
OrderUUID is generated by Openfinex when we forward the new Order struct, so how do we propose to store it in orderbook storage? also, waiting for the result from openfinex might slow down the enclave since it is a network call, even though it is localhost

Reply:
good point. We could cache the order until we have a uuid from openfinex. But then how would we match return values with cached orders?

Reply:
Hello, when we were solving a similar case we got this algorithm:

1. Websocket API returns back the number you put in response, so when I'm submitting an order a generate incremental number and cache it.
 For request
```json
[1, 48, "method", [...args]]
```
Response will be 
```json
[2(means response), 48(my number sent back), "method", [..info]]
```
2. API returns me my number and generated uuid, so I put uuid into a cache.
3. After engine creates an order I'll get a notification with uuid I have in cahce

fyi

Reply:
> Hello, when we were solving a similar case we got this algorithm:
> 
> 1. Websocket API returns back the number you put in response, so when I'm submitting an order a generate incremental number and cache it.
>    For request
> 
> ```json
> [1, 48, "method", [...args]]
> ```
> 
> Response will be
> 
> ```json
> [2(means response), 48(my number sent back), "method", [..info]]
> ```
> 
> 1. API returns me my number and generated uuid, so I put uuid into a cache.
> 2. After engine creates an order I'll get a notification with uuid I have in cahce
> 
> fyi

Thanks, Denis. This solution works for us!